stan/math/opencl/matrix_cl.hpp

```diff
@@ -341,6 +347,47 @@ class matrix_cl<T, enable_if_arithmetic<T>> {
     }
   }
 
+  /**
+   * Constructs a const matrix_cl that contains a copy of the Eigen matrix on
+   * the OpenCL device. If the matrix already has a cached copy on the device,
+   * the cache is used and no copying is done. Changing the resulting matrix_cl
+   * would change cache, so do not do it! If changes are needed a copy must be
+   * made.
+   *
+   * @tparam R row type of input matrix
+   * @tparam C column type of input matrix
+   * @param A the Eigen matrix
+   * @param partial_view which part of the matrix is used
+   */
+  template <int R, int C>
+  static matrix_cl<T> constant(const Eigen::Matrix<T, R, C>& A,
```

Comment:
Why is this static? And why don't we just have a template flag  for `bool cache = false` instead of a new member function?

Reply:
If it was a member function instead of static one, we would need a matrix_cl object to call it on. To construct a new object it needs to be static.

It is impossible to explicitely set a constructor template parameter (One can set class template parameters. Constructor template parameters can only be deduced.)

Reply:
Adding a template to class just because we need two ways of constructing it from same argument type seems like an overkill. Especially since those two template instantinations of matrix_cl are identical once constructed. That would mean two versions or an additional template on every function that uses matrix_cl. Having separate static function for constructing one version is much simpler.

> Counterpoint to that though is if we will ever want to know in the future if something was cached or not.

Can you think of example where this would be needed?

Reply:
Not at the moment so I'm cool with the way we are doing this. Though I do want to think about that for a bit. But I also like that your way is v explicit